packages/core/src/modules/vc/W3cV2CredentialServiceOptions.ts

```diff
@@ -0,0 +1,229 @@
+import { KnownJwaSignatureAlgorithm } from '../kms'
+import { IDisclosureFrame, IPresentationFrame, SdJwtVcHolderBinding } from '../sd-jwt-vc'
+import { W3cV2JwtVerifiableCredential, W3cV2JwtVerifiablePresentation } from './jwt-vc'
+import type { ClaimFormat, W3cV2VerifiableCredential } from './models'
+import type { W3cV2Credential, W3cV2Presentation } from './models'
+import { W3cV2SdJwtVerifiableCredential, W3cV2SdJwtVerifiablePresentation } from './sd-jwt-vc'
+
+export type W3cV2SignCredentialOptions<
+  Format extends ClaimFormat.JwtW3cVc | ClaimFormat.SdJwtW3cVc | undefined = undefined,
+> = Format extends ClaimFormat.JwtW3cVc
+  ? W3cV2JwtSignCredentialOptions
+  : Format extends ClaimFormat.SdJwtW3cVc
+    ? W3cV2SdJwtSignCredentialOptions
+    : W3cV2JwtSignCredentialOptions | W3cV2SdJwtSignCredentialOptions
+
+export type W3cV2VerifyCredentialOptions<
+  Format extends ClaimFormat.JwtW3cVc | ClaimFormat.SdJwtW3cVc | undefined = undefined,
+> = Format extends ClaimFormat.JwtW3cVc
+  ? W3cV2JwtVerifyCredentialOptions
+  : Format extends ClaimFormat.SdJwtW3cVc
+    ? W3cV2SdJwtVerifyCredentialOptions
+    : W3cV2JwtVerifyCredentialOptions | W3cV2SdJwtVerifyCredentialOptions
+
+export type W3cV2SignPresentationOptions<
+  Format extends ClaimFormat.JwtW3cVp | ClaimFormat.SdJwtW3cVp | undefined = undefined,
+> = Format extends ClaimFormat.JwtW3cVp
+  ? W3cV2JwtSignPresentationOptions
+  : Format extends ClaimFormat.SdJwtW3cVp
+    ? W3cV2SdJwtSignPresentationOptions
+    : W3cV2JwtSignPresentationOptions | W3cV2SdJwtSignPresentationOptions
+
+export type W3cV2VerifyPresentationOptions = W3cV2JwtVerifyPresentationOptions | W3cV2SdJwtVerifyPresentationOptions
+
+interface W3cV2SignCredentialOptionsBase {
+  /**
+   * The format of the credential to be signed.
+   *
+   * @see https://identity.foundation/claim-format-registry
+   */
+  format: ClaimFormat
+
+  /**
+   * The credential to be signed.
+   */
+  credential: W3cV2Credential
+
+  /**
+   * URI of the verificationMethod to be used for signing the credential.
+   *
+   * Must be a valid did url pointing to a key.
+   */
+  verificationMethod: string
+}
+
+export interface W3cV2JwtSignCredentialOptions extends W3cV2SignCredentialOptionsBase {
+  format: ClaimFormat.JwtW3cVc
+
+  /**
+   * The alg to be used for signing the credential.
+   *
+   * Must be a valid JWA signature algorithm.
+   */
+  alg: KnownJwaSignatureAlgorithm
+}
+
+export interface W3cV2SdJwtSignCredentialOptions extends W3cV2SignCredentialOptionsBase {
+  format: ClaimFormat.SdJwtW3cVc
+
+  /**
+   * The algorithm used to sign the SD-JWT VC.
+   */
+  alg: KnownJwaSignatureAlgorithm
+
+  /**
+   * The hashing algorithm to use for creating digests of the disclosures.
+   *
+   * @default 'sha-256'
+   */
+  hashingAlgorithm?: Exclude<HashName, 'sha-1'>
+
+  /**
+   * The holder to bind the SD-JWT VC to. If no holder is provided, the SD-JWT
+   * VC won't be bound to the holder.
+   */
+  holder?: SdJwtVcHolderBinding
```

Comment:
I think we should not reuse the `SdJwtVcHolderBinding` here, since that also allows `jwk` binding. I'm not sure how that will work combined with `credntialSubject.id`? For W3C v1 we only support dids and i think it might make sense to also do that for v2? 

Currently i think you have to set the `credentialSubject.id` in the credential, and there's no `holder` property.

We do check that the `cnf` and `credentialSubject.id` match, so what would happen if we provide a jwk here?

Reply:
The usage of `cnf` seems to be recommended: https://www.w3.org/TR/vc-jose-cose/#cnf

What if we:

- Remove `holder` and make it `holderBinding?: boolean`
- When that is `true`, make a holder binding based on the first DID in `     options.credential.credentialSubjectIds`

?